### ISSUE TITLE
Unexpected type warning with Click's callback

### ISSUE THREAD

**Issue by jaraqueffdc** [microsoft/pylance-release#838]
Hi, I haven't been able to find anything about this issue and since it is very specific it might not be common but pylance y marking a call to `@click.option` when I do not see a reason why that would be the case.

This is the example. DataStore is just an enum we use for our DataStore

```python
def parse_dstore_param(
    ctx: click.Context,
    param: click.Option,
    value: str,
) -> DataStore:
    return DataStore[value.upper()]


@click.group(no_args_is_help=False, invoke_without_command=True)
@click.option("--verbose", default=False, is_flag=True)
@click.option(
    "--data-store",
    help="Data store to use for execution byproducts",
    show_default=True,
    default=DataStore.LOCAL.value,
    type=click.Choice([a.name.lower() for a in DataStore], case_sensitive=False),
    callback=parse_dstore_param,
    is_eager=True,
)
```

As you can see, the callback method is a `Callable[[click.Context, click.Option, str], DataStore]`. This should match the stub for it defined like this:

```python
def option(
    *param_decls: str,
    cls: Type[Option] = ...,
    # Option
    show_default: Union[bool, Text] = ...,
    prompt: Union[bool, Text] = ...,
    confirmation_prompt: bool = ...,
    hide_input: bool = ...,
    is_flag: Optional[bool] = ...,
    flag_value: Optional[Any] = ...,
    multiple: bool = ...,
    count: bool = ...,
    allow_from_autoenv: bool = ...,
    type: Type[str] = ...,
    help: Optional[str] = ...,
    show_choices: bool = ...,
    # Parameter
    default: Optional[Any] = ...,
    required: bool = ...,
    callback: Callable[[Context, Union[Option, Parameter], str], Any] = ...,
    nargs: Optional[int] = ...,
    metavar: Optional[str] = ...,
    expose_value: bool = ...,
    is_eager: bool = ...,
    envvar: Optional[Union[str, List[str]]] = ...,
    # User-defined
    **kwargs: Any,
) -> _IdentityFunction: ...
```

That defines the callback as `Callable[[Context, Union[Option, Parameter], str], Any]` but for some reason, the only way I can get it not to show the error is if I take the `click.Option` type annotation out.

Do you guys have an idea of the issue might be? I am marking this as a question because I think it is likely I am missing something but I am not sure if this is actually a bug.

**Comment by judej** [comment]
@erictraut, would you please look at this?

**Comment by erictraut** [comment]
The problem is with the second parameter (`param`) in the function `parse_dstore_param`. It is annotated as accepting a value of type `Option`, but the `callback` parameter requires a function that accepts a value that is of type `Union[Option, Parameter]`. If the callback were to pass a `Parameter` value to the callback, it would not work. The error message is unfortunately not so clear because `click.option` uses overloads, and there's no great way to report precise error messages for overloads.

So, in summary, pylance is doing the right thing here. It's telling you that `parse_dstore_param` has an incompatible signature. You can fix it by changing it to:
```python
def parse_dstore_param(
    ctx: click.Context,
    param: Union[click.Option, click.Parameter],
    value: str,
) -> DataStore:
    return DataStore[value.upper()]
```

**Comment by jaraqueffdc** [comment]
For some reason I thought that as long as my type was coherent with the Union type it would work. I even get mypy not complaining though, it was just pylance the one that was complaining about it. In any case, thanks a lot! That fixes it and now everything is clean again :)

**Comment by erictraut** [comment]
Input parameter types for callables are considered "contravariant", which means that the dest type needs to be assignable to the source type — the opposite of what you are normally used to. In your case, `Union[Option, Parameter]` was not assignable to `Option`, hence the error. It might seem counterintuitive at first, but it makes sense when you think about it from the perspective of the code that will call the callback function.

**Comment by jaraqueffdc** [comment]
Yes, it makes complete sense, I just was unaware of it. Thanks a lot!
